tests/unit/conftest.py

```diff
@@ -0,0 +1,155 @@
+        # cfg['letsencrypt-domains'] = 'mock'
+        return cfg
+
+    monkeypatch.setattr("libwireguard.hookenv.config", mock_config)
+
+
+@pytest.fixture
+def mock_action_set(monkeypatch):
+    """Mock action_set to facilitate testing of action results."""
+    mocked_action_set = mock.Mock(returnvalue=True)
+    monkeypatch.setattr("charmhelpers.core.hookenv.action_set", mocked_action_set)
+    return mocked_action_set
+
+
+@pytest.fixture
+def mock_opened_ports(monkeypatch):
+    """Mock the charmhelpers hookenv list of open ports."""
+    def mocked_opened_ports_sa():
+        return ['51902/udp', '11111/tcp', '22222/udp']
+
+    mocked_opened_ports = mock.Mock()
+    mocked_opened_ports.side_effect = mocked_opened_ports_sa
+    monkeypatch.setattr('libwireguard.hookenv.opened_ports', mocked_opened_ports)
+    return mocked_opened_ports
+
+
+@pytest.fixture
+def mock_open_port(monkeypatch):
+    """Mock the open port call."""
+    mocked_open_port = mock.Mock()
+    monkeypatch.setattr('libwireguard.hookenv.open_port', mocked_open_port)
+    return mocked_open_port
+
+
+@pytest.fixture
+def mock_close_port(monkeypatch):
+    """Mock the close port call."""
+    mocked_close_port = mock.Mock()
+    monkeypatch.setattr('libwireguard.hookenv.close_port', mocked_close_port)
+    return mocked_close_port
+
+
+@pytest.fixture
+def mock_service(monkeypatch):
+    """Mock charmhelpers service control."""
+    mocked_service = mock.Mock()
+    monkeypatch.setattr('libwireguard.host.service', mocked_service)
+    return mocked_service
+
+
+@pytest.fixture
+def mock_charm_dir(monkeypatch):
+    """Mock charm working directory."""
+    monkeypatch.setattr("libwireguard.hookenv.charm_dir", lambda: ".")
+
+
+@pytest.fixture
+def mock_render(monkeypatch):
+    """Mock templating render function."""
+    from functools import partial
+    from charmhelpers.core import templating
+    import getpass
+
+    user = getpass.getuser()
+
+    monkeypatch.setattr(
+        "libwireguard.templating.render",
+        partial(templating.render, owner=user, group=user),
+    )
+
+
+@pytest.fixture
+def mock_subprocess_check_call(monkeypatch):
+    """Mock calls to the check_call function."""
+    mocked_check_call = mock.Mock()
+    monkeypatch.setattr("libwireguard.check_call", mocked_check_call)
+    return mocked_check_call
+
+
+@pytest.fixture
+def mock_subprocess_popen(monkeypatch):
+    """Mock calls to the Popen function."""
+    class MockedClassPopen():
+
+        def __init__(self, args, stdin=None, stdout=None, stderr=None):
+            return None
+
+        def communicate(self, byteinput):
+            return [b"mocked-stdout", b"mocked-stderr"]
+
+    def mocked_sa_popen(cmd, stdin=None, stdout=None, stderr=None):
+        return MockedClassPopen(cmd, stdin, stdout, stderr)
+
+    mocked_popen = mock.Mock()
+    mocked_popen.side_effect = mocked_sa_popen
+    monkeypatch.setattr("libwireguard.Popen", mocked_popen)
+    return mocked_popen
+
```

Comment:
I notice that the unit-state.db was added to the .gitignore. If the unit tests are generating that file it would be best to mock it to avoid the file creation. I can't think of any downside, this should just go into the template now that I've learned this is so easy to do.

```python
from charmhelpers.core import unitdata

@pytest.fixture
def mock_unit_db(monkeypatch):
    """Mock the key value store."""
    mock_kv = mock.Mock()
    mock_kv.return_value = unitdata.Storage(path=":memory:")
    monkeypatch.setattr("libwireguard.unitdata.kv", mock_kv)
```
Then add mock_unit_db to the wh below and unit testing shouldn't generate the unit-state.db file anymore.